### PR TITLE
Update RURQ -- Added Editorial

### PR DESCRIPTION
**Pull request by Qw3rTy827**
I have added an editorial to the problem "Range Updates & Sums"

### CODE REVIEW

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:246]: > -Range

Did you check this in liveupdate?

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:381]: `endl` may flush the output, use `\n`

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:253]: not a fan of this

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:244]: All sentences should end with periods.

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:236]: would prefer more informative variable names (e.g. `lz_set, lz_add`) even though it makes the code longer

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:248]: add backticks around pushdown like so: `pushdown`

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:242]: > doesnt equal to 0

"is not equal to 0" or "doesn't equal 0" or $lz2\neq 0$ preferred

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx:246]: Yes I have checked this in liveupdate. The unordered list function works fine on my side.

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx:381]: Problem addressed.

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx:253]: Problem addressed.

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx:244]: Problem addressed.

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx:242]: Addressed.

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx:246]: Typo on my part. Addressed.

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx:236]: I changed lz1 and lz2 to lzAdd and lzSet respectively.

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx:248]: Addressed.

**Comment by Qw3rTy827** [comment]
Btw, just for future reference, do I open a new pull request or use the old one (like I have done so this time) if I need to address errors with my initial pull request. (I am new to GitHub).

**Comment by thecodingwizard** [comment]
Use the new one :) Thanks for contributing!

**Comment by Qw3rTy827** [comment]
> Use the old one -- what you did this time is perfect :) Thanks for contributing!

No problem !

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [reply content/5_Plat/RURQ.mdx:253]: This was not changed.

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [reply content/5_Plat/RURQ.mdx]: Not addressed.

**Review by bqi343 — CHANGES_REQUESTED** [review]
address comments above, also use tabs rather than spaces

**Comment by Qw3rTy827** [comment]
> address comments above, also use tabs rather than spaces

Oops. I might have accidentally pushed in an older version of the mdx file. It should be fine now. Sorry I am still experimenting with Github's functions :/

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx]: Should be good now.

**Comment by dongliu0426** [comment]
> also use tabs rather than spaces

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [reply content/5_Plat/RURQ.mdx]: Not addressed.

**Review by bqi343 — CHANGES_REQUESTED** [review]
see @dongliu0426 's comment and my comment above

**Comment by Qw3rTy827** [comment]
> > also use tabs rather than spaces

Do you mean tabs rather than spaces after items in lists ?

**Comment by bqi343** [comment]
in the code snippet

**Comment by Qw3rTy827** [comment]
> in the code snippet

indents in the code section ?

**Comment by Qw3rTy827** [comment]
> snippet

Is there an efficient to change the spaces to tabs. I copied my code from my IDE (Clion) and it turned out to be what it is right now.

**Comment by bqi343** [comment]
any IDE should have an option to convert indentation

**Comment by Qw3rTy827** [comment]
> any IDE should have an option to convert indentation

Ok. I will try that right now.

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:242]: > doesn't equal to 0

still hasn't been changed

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx:242]: > > doesn't equal to 0
> 
> still hasn't been changed

Didn't u say that "doesn't equal to 0" is ok though? What should I change it to if this is not okay ?

Comment (bqi343) [reply content/5_Plat/RURQ.mdx:242]: "doesn't equal 0" was ok, "doesn't equal to 0" is not

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:253]: > #define int long long

why is this still here :/

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Comment by thecodingwizard** [comment]
To fix build either merge master into this branch or just ignore the failing build and merge straight into master ;)

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx]: I replaced it with #define ll long long

**Comment by Qw3rTy827** [comment]
> To fix build either merge master into this branch or just ignore the failing build and merge straight into master ;) (it should build once merged)

Um I do not have the authority to merge. It says that "Merging is blocked" for me.

**Comment by thecodingwizard** [comment]
yep that's intentional, no worries. Once Ben is done reviewing he will merge (and/or ask for another reviewer)

**Comment by Qw3rTy827** [comment]
> yep that's intentional, no worries. Once Ben is done reviewing he will merge (and/or ask for another reviewer)

Understand. So do I just wait for Ben's review right now ?

**Comment by HarryWangATX** [comment]
yes

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [reply content/5_Plat/RURQ.mdx]: As mentioned [here](https://usaco.guide/general/generic?lang=cpp#macros), `using ll = long long;` is preferred.

Also, please don't just make everything `ll` ...

Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx]: Fixed. On a side note would making everything long long be considered as a bad habit that I should avoid if I am sure that MLE is not going to be an issue for this particular problem.

Comment (bqi343) [reply content/5_Plat/RURQ.mdx]: well I guess it isn't that bad, but if you use different data types for the indices vs the values it's easy to see what you need to change if you want to use a different value type.

btw - the code now fails ...
Screen Shot 2021-03-30 at 10 50 12 AM

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:353]: since you store `l` and `r` in `tree[p]`, you don't need to pass `l` and `r` as arguments to this function

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx]: yea I realized that I overchanged long long variables back to int and caused overflow... I think thats how my habit of making all integers long long started. I tested again and it is AC this time.

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx:353]: True. Should I change this or is this just for future reference ?

Comment (bqi343) [reply content/5_Plat/RURQ.mdx:353]: change

Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx]: I removed ```l``` and ```r``` in each node of the tree to minimize unnecessary redundancy.

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:316]: `l` and `r` can be `int`s

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:319]: this `1ll` at the end does nothing

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx]: Unnecessary ```1ll```s removed.

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [reply content/5_Plat/RURQ.mdx:316]: ^ address this. Only `val` needs to be `ll`

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx:325]: Fixed.

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:332]: same for l,r here ...

File: content/5_Plat/RURQ.mdx
Comment (bqi343) [inline content/5_Plat/RURQ.mdx:303]: and here and above

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx]: Updated.

Comment (bqi343) [reply content/5_Plat/RURQ.mdx]: what about in `pushdown`?

Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx]: Changed.

**Review by bqi343 — APPROVED** [review]

File: content/5_Plat/RURQ.mdx
Comment (dolphingarlic) [inline content/5_Plat/RURQ.mdx:228]:
```suggestion
This question asks you to support the following types of queries:
```

File: content/5_Plat/RURQ.mdx
Comment (dolphingarlic) [inline content/5_Plat/RURQ.mdx:230]: Change these tabs to single spaces for all list items

File: content/5_Plat/RURQ.mdx
Comment (dolphingarlic) [inline content/5_Plat/RURQ.mdx:236]:
```suggestion
Consider the first two types of queries. A lazy tag will be created in each node of the tree for each type. In this solution, $lzAdd$ will represent the lazy tag for the range add query and $lzSet$ will represent the lazy tag for the range set query. 
```

Also, it might look better if you do something like $\texttt{lzAdd}$ instead of just $lzAdd$, but either one is fine with me

File: content/5_Plat/RURQ.mdx
Comment (dolphingarlic) [inline content/5_Plat/RURQ.mdx:238]:
```suggestion
Given the two different types of update queries, a total of four different situations might take place after any update:
```

File: content/5_Plat/RURQ.mdx
Comment (dolphingarlic) [inline content/5_Plat/RURQ.mdx:246]:
```suggestion
-	Range set when $lzAdd$ doesn't equal 0: Again, simply update the $lzSet$ value since a set update will override all previous add updates.
```

File: content/5_Plat/RURQ.mdx
Comment (dolphingarlic) [inline content/5_Plat/RURQ.mdx:248]:
```suggestion
Given the mechanics behind the `pushdown` function, all you need is a regular range-sum segment tree to solve the question. 
```

**Review by dolphingarlic — CHANGES_REQUESTED** [review]
Just some minor style adjustments

File: content/5_Plat/RURQ.mdx
Comment (Qw3rTy827) [reply content/5_Plat/RURQ.mdx]: All suggested changes are applied.

**Pull request merged**
